### DISCUSSION TITLE
Serving static files over http from node.js

### DISCUSSION THREAD

**Discussion by hcldan** [nodejs/help#4905] [Q&A]
I have seen numerous recommendations online for running NGINX as a caching RP in front of node, so that node doesn't have to server static files... because it is bad at it, but never really an explanation as to *why*...

Some of the older posts mention things like readFile in express-static (which now pipes the stream), but most posts just say "use nginx".
I'd really like to understand why, and to understand the metrics I'm seeing in my current benchmark.

I am seeing relatively low cpu/memory/network utilization, (8 core, 16G mem EC2 instance) with a simple static server using cluster to use n-1 forks and serving up pre-compressed gzip files from the filesystem.

I'm at a loss for where the bottleneck is here...  any tips?

**Comment by Trott** [comment]
While I don't doubt that there are performance issues, I always thought the bigger issue was security. Things like `nginx` and Apache `httpd` are battle-tested and designed to handle hostile input. When they fail, they fail and restart gracefully. Node.js is not built with these things as a design consideration. As such, I would not recommend exposing it directly as an http server for all the same reasons you wouldn't expose a Python script as an http server. I'd put it behind `nginx` or Apache `httpd` or something similar to that.

**Comment by jasnell** [comment]
Performance really is the biggest issue with regards to serving static files from Node.js. The bottleneck is largely due to the way Node.js reads and flows data through the system.

Consider this: When you use the async file system APIs to read a file, each individual read is deferred to the libuv thread pool, queued by the event loop for processing, marshalled across the C++/JavaScript boundary, passed into JavaScript operations to possibly be manipulated by user code, then passed back out to the http socket which marshals each chunk back across the C++/JavaScript boundary and back down to libuv for sending. It's a significant amount of additional overhead. If we happen to layer TLS into that flow as well it just becomes more expensive. `nginx` and `httpd` handle that flow much more efficiently overall.

**Reply by hcldan** [reply]
> The bottleneck is largely due to the way Node.js reads and flows data through the system.

Now this is something I've found on early versions of node/express where the response bits were shuffled into a buffer and back out to the network.   I'm actually pretty impressed with the performance now with serving static files... has this concern been addressed or is there still a large amount of marshalling with the stream piping?

This is a much more detailed response than anything I've been able to find, thank you for that.  I hope it helps others wading through all the old nodejs 0.x posts :)
